primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -0,0 +1,635 @@
+	is removed its containing dirty set is removed. This function is only called for keys that
+	are in the dirty set. Therefore the entry must exist; qed";
+
+const PROOF_OVERLAY_NON_EMPTY: &str = "\
+	An OverlayValue is always created with at least one transaction and dropped as soon
+	as the last transaction is removed; qed";
+
+type DirtyKeys = SmallVec<[HashSet<StorageKey>; 5]>;
+type Transactions = SmallVec<[InnerValue; 5]>;
+
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+struct InnerValue {
+	/// Current value. None if value has been deleted.
+	value: Option<StorageValue>,
+	/// The set of extrinsic indices where the values has been changed.
+	/// Is filled only if runtime has announced changes trie support.
+	extrinsics: BTreeSet<u32>,
+}
+
+/// An overlay that contains all versions of a value for a specific key.
+#[derive(Debug, Default, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct OverlayedValue {
+	// The individual versions of that value.
+	// One entry per transactions during that the value was actually written.
+	transactions: Transactions,
+}
+
+/// Holds a set of changes with the ability modify them using nested transactions.
+#[derive(Debug, Default, Clone)]
+pub struct OverlayedChangeSet {
+	/// Stores the changes that this overlay constitutes.
+	changes: BTreeMap<StorageKey, OverlayedValue>,
+	/// Stores which keys are dirty per transaction. Needed in order to determine which
+	/// values to merge into the parent transaction on commit. The length of this vector
+	/// therefore determines how many nested transactions are currently open (depth).
+	dirty_keys: DirtyKeys,
+}
+
+impl OverlayedValue {
+	/// The value as seen by the current transaction.
+	pub fn value(&self) -> Option<&StorageValue> {
+		self.transactions.last().expect(PROOF_OVERLAY_NON_EMPTY).value.as_ref()
+	}
+
+	/// Unique list of extrinsic indices which modified the value.
+	pub fn extrinsics(&self) -> impl Iterator<Item=&u32> {
+		self.transactions.iter().flat_map(|t| t.extrinsics.iter()).unique()
+	}
+
+	// Mutable reference to the most recent version.
+	fn value_mut(&mut self) -> &mut Option<StorageValue> {
+		&mut self.transactions.last_mut().expect(PROOF_OVERLAY_NON_EMPTY).value
+	}
+
+	// Remove the last version and return it.
+	fn pop_transaction(&mut self) -> InnerValue {
+		self.transactions.pop().expect(PROOF_OVERLAY_NON_EMPTY)
+	}
+
+	// Mutable reference to the set which holds the indices for the **current transaction only**.
+	fn transaction_extrinsics_mut(&mut self) -> &mut BTreeSet<u32> {
+		&mut self.transactions.last_mut().expect(PROOF_OVERLAY_NON_EMPTY).extrinsics
+	}
+
+	// Writes a new version of a value.
+	//
+	// This makes sure that the old version is not overwritten and can be properly
+	// rolled back when required.
+	fn set(
+		&mut self,
+		value: Option<StorageValue>,
+		first_write_in_tx: bool,
+		at_extrinsic: Option<u32>
+	) {
+		if first_write_in_tx || self.transactions.is_empty() {
```

Comment:
No. I need to push it because the value was already marked as dirty. Not pushing it would result in an imbalance at transaction close.

Reply:
You just need to not mark it as dirty then.

Reply:
It is not 'just'. I have to add more conditionals to at least two callers and think about more corner cases. It complicates the code further for no clear benefit. Keep in mind to determine equality we have to read each and every value passed here. Depending on the length it can be even slower. For that reason I am closing this as currently I am mainly concerned with correctness.

Reply:
Yes it is not false we got those values that serialize some whole array and could only change on the very last bits. The main point here is to avoid some memory allocation.
Shouldn't have use 'just' word, but seems like we agree that this suggestion will not result in an imbalance and is only a matter of adding two conditionals.
Correctness is ok btw.